spring-integration-core/src/main/java/org/springframework/integration/handler/DelayHandler.java

```diff
@@ -370,11 +457,75 @@ private void releaseMessageAfterDelay(final Message<?> message, long delay) {
 	}
 
 	private void releaseMessage(Message<?> message) {
-		this.releaseHandler.handleMessage(message);
+		String identity = ObjectUtils.getIdentityHexString(message);
+		this.deliveries.putIfAbsent(identity, new AtomicInteger());
+		try {
+			this.releaseHandler.handleMessage(message);
+			this.deliveries.remove(identity);
+		}
+		catch (Exception e) {
+			if (getErrorChannel() != null) {
+				ErrorMessage errorMessage = new ErrorMessage(e,
+						Collections.singletonMap(IntegrationMessageHeaderAccessor.DELIVERY_ATTEMPT,
+								new AtomicInteger(this.deliveries.get(identity).get() + 1)),
+						message);
+				try {
+					if (!(getErrorChannel().send(errorMessage))) {
+						if (this.logger.isDebugEnabled()) {
+							this.logger.debug("Failed to send error message: " + errorMessage);
+						}
+						rescheduleForRetry(message, identity);
+					}
+					else {
+						this.deliveries.remove(identity);
+					}
+				}
+				catch (Exception e1) {
+					if (this.logger.isDebugEnabled()) {
+						logger.debug("Error flow threw an exception for message: " + message, e1);
+					}
+					rescheduleForRetry(message, identity);
+				}
+			}
+			else {
+				if (this.logger.isDebugEnabled()) {
+					logger.debug("Release flow threw an exception for message: " + message, e);
+				}
+				if (!rescheduleForRetry(message, identity)) {
+					throw e; // there might be an error handler on the scheduler
```

Comment:
Why do we rethrow exception only here?
How about all other `rescheduleForRetry()` usage?

Reply:
This code was added to support "legacy" error channel processing; the new error channel processing is performed within the scope of the transaction (if present); the old one is outside the transaction scope. There are several tests that use the `MessagePublishingErrorHandler` in the scheduler.

The legacy error handling is unchanged in that it is only invoked after the final release attempt. It also cannot influence the retry logic (unlike the new version).